### PR TITLE
B56 add refund

### LINKED ISSUE #291: B.5.6 Add Refund Request

### PR DESCRIPTION
**Pull request by ianleck**
## Description:
- Added `RefundRequest`
- Edited billing to include `mentorPassCount`
- Added `refunded` enum for `BILLING_STATUS`
- Added request refund, approve/reject still needs some work so the routes for those are not up yet. Making the PR first to limit the PR size.
## Checklist:
- [x] Merged latest develop
- [x] PR title makes sense

### CODE REVIEW

File: src/services/paypal.service.ts [hunk 1]
```diff
@@ -81,7 +66,254 @@ export default class PaypalService {
...
+    if (!paymentId) throw new Error(WALLET_ERROR.INVALID_REFUND_REQUEST);
+
+    const refund_details = {
+      amount: {
+        currency: billing.currency,
+        total: `${billing.amount}`,
+      },
+    };
+    return { paymentId, refund_details };
+  }
```
Comment (ianleck) [inline src/services/paypal.service.ts]: Ignore for now

File: src/services/paypal.service.ts [hunk 2]
```diff
@@ -81,7 +66,254 @@ export default class PaypalService {
...
+        billingType: BILLING_TYPE.MENTORSHIP,
+      },
+      order: ['createdAt', 'DESC'],
+    });
+
+    let billingIds: string[] = [];
+    let remainingNumPasses = mentorshipContract.mentorPassCount;
+    for (let i = 0; i < paidBillings.length; i++) {
+      if (remainingNumPasses <= 0) break;
+      const billing = paidBillings[0];
```
Comment (kengthong) [inline src/services/paypal.service.ts]: This looks like a bug right? shld it be const billing = paidBillings[i]

File: src/services/paypal.service.ts [hunk 3]
```diff
@@ -81,7 +66,254 @@ export default class PaypalService {
...
+      include: [
+        {
+          model: Billing,
+          as: 'OriginalBilling',
+          where: { billingId: billing.billingId },
+          include: [{ model: Course, as: 'Course' }],
+        },
+      ],
+    });
+  }
```
Comment (kengthong) [inline src/services/paypal.service.ts]: you didnt include billingId when creating requestRefund but u can query with where billingId in findOne?

File: src/models/MentorshipContract.ts [hunk 4]
```diff
@@ -46,10 +43,10 @@ export class MentorshipContract extends BaseEntity {
   @Column({
     allowNull: false,
     type: DataType.ENUM,
-    values: Object.values(MENTORSHIP_CONTRACT_APPROVAL),
-    defaultValue: MENTORSHIP_CONTRACT_APPROVAL.PENDING,
+    values: Object.values(APPROVAL_STATUS),
+    defaultValue: APPROVAL_STATUS.PENDING,
   })
-  senseiApproval: MENTORSHIP_CONTRACT_APPROVAL;
+  senseiApproval: APPROVAL_STATUS;
```
Comment (kengthong) [inline src/models/MentorshipContract.ts:50]: Do you need to add @AfterDestroy to delete related consultations and tasks?

**Review by kengthong — CHANGES_REQUESTED** [review]

File: src/models/MentorshipContract.ts (hunk 4 above)
Comment (ianleck) [reply src/models/MentorshipContract.ts:50]: Not destroying mentorship aft rejection. Anw rejection not done

File: src/services/paypal.service.ts [hunk 5]
```diff
@@ -81,7 +66,254 @@ export default class PaypalService {
...
 
-  public static async createRefund() {}
+  // ============================== Refunds ==============================
+  // =============== Request Refunds ===============
+  public static async requestRefund(
+    contractId: string,
+    contractType: string,
+    accountId: string
+  ) {
+    try {
```
Comment (y1bs) [inline src/services/paypal.service.ts:77]: Do you need the billingId for this too?

File: src/services/paypal.service.ts (hunk 3 above)
Comment (ianleck) [reply src/services/paypal.service.ts]: Ya can but need use on I've updated it

File: src/services/paypal.service.ts (hunk 5 above)
Comment (ianleck) [reply src/services/paypal.service.ts:77]: Don't need because if by billingId then mentorpass side will not work. I originally tried with billingId

Comment (y1bs) [reply src/services/paypal.service.ts:77]: Ahh okok

**Review by y1bs — APPROVED** [review]
Looks ok from the frontend perspective, thanks!

File: src/models/MentorshipContract.ts (hunk 4 above)
Comment (kengthong) [reply src/models/MentorshipContract.ts:50]: I was referring to afterDestroy when a mentorshipContract is destroyed, added a new comment bout it

File: src/services/paypal.service.ts [hunk 6]
```diff
@@ -81,7 +67,283 @@ export default class PaypalService {
...
+
+    // 3. Destroy CouseContract/MentorshipContract
+    if (originalBilling.billingType === 'COURSE') {
+      await CourseContract.destroy({
+        where: { courseContractId: originalBilling.contractId },
+      });
+    } else {
+      await MentorshipContract.destroy({
+        where: { mentorshipContractId: originalBilling.contractId },
+      });
```
Comment (kengthong) [inline src/services/paypal.service.ts:308]: The afterDestroy comment was actually about this sorry

**Review by kengthong — APPROVED** [review]

**Pull request merged**